src/ast/semantic_analyser.cpp

```diff
@@ -1704,5 +1705,32 @@ void SemanticAnalyser::assign_map_type(const Map &map, const SizedType &type)
   }
 }
 
+bool SemanticAnalyser::is_context_access(Expression *expr, bool is_pointer_dereferenced)
+{
+  // Example:
+  //   1. args->filename;
+  //   2. ((struct perf_event_data*)ctx)->regs.ip;
+  //   3. $a = ((struct perf_event_data*)ctx); $a->sample_period;
+
+  if (auto *p = dynamic_cast<Builtin*>(expr))
+    return p->ident == "ctx" || p->ident == "args";
+  else if (auto *p = dynamic_cast<Variable*>(expr))
+    return is_context_access(variable_[p->ident], is_pointer_dereferenced);
+  else if (auto *p = dynamic_cast<Binop*>(expr))
```

Comment:
in which case would this be hit?

Reply:
For example:

```
$a = ((struct perf_event_data*)ctx);
@ = $a->sample_period;
```

AST is

```
Program
 software:cpu:1000000000
  =
   variable: $a
   (bpf_perf_event_data*)
    builtin: ctx
  =
   map: @
   .
    dereference
     variable: $a
    sample_period
```

To see if `$a` is ctx or not it is necessary to check the expression assigned to the variable `$a`.
`Variable` class does not have a `expr` member (`AssignmentVarStatement` does.)

Reply:
Yes variables are needed, but the assignment is not a binop. I can't find a case in which this binop case would be required. E.g. with this patch:

```diff
+  else if (auto *p = dynamic_cast<Binop*>(expr)) {
+    std::cout << "BINOP CONTEXT" << std::endl;
     return is_context_access(p->left, is_pointer_dereferenced) ||
              is_context_access(p->right, is_pointer_dereferenced);
+  }
```

and this program:
```
include "linux/bpf_perf_event.h"

software:cpu:1000000000 {
  $a = ((struct bpf_perf_event_data*)ctx);
  @ = $a->sample_period;
  @a = $a->sample_period * $a->regs.ip;
  @a += $a->regs.ax;
  $b = $a->sample_period * $a->regs.ip;
  $b *= $b;
  @c = $b + @;
}
```

I still don't hit the binop path.

Reply:
Ah, I misunderstood your question.

I considered something like the followiong. 
```
@ = *(ctx + 40);
```

There is no incentive to do this generally but it is possible... but now it turns out this does not work as expected because this patch currently only considers dereference in the FieldAccess and does not generate a load instruction for Unop dereference. I think this can be fixed by doing the same thing for Unop as does for FieldAccess. If there is no objection, I'll work on to support unop dereference.

Reply:
FYI, the following is possible.

```
struct perfdata
{
  __u64 sample_period;
  __u64 addr;
}

software:cpu:1000000000
{
  $perfdata = (struct perfdata*)(ctx + 168);
  printf("%lx\n", $perfdata->sample_period);
}
```

Reply:
Ah yes that case would trigger it. It doesn't seem like anything you'd want to use in real life. What do we gain by supporting it (except more complexity)? Telling the users it is unsupported (it is already a hidden builtin) feels like a better option for this.

Reply:
OK, print error messages then.